lib/phoenix/presence.ex

```diff
@@ -421,4 +430,64 @@ defmodule Phoenix.Presence do
       end)
     end)
   end
+
+  @doc """
+  Keep a presence state in sync.
+    * `presence_list` - A list of presences that was obtained by calling `Presence.list/1`
+    * `diff` - The `payload` from a `presence_diff` broadcast.
+  ## Examples
+      iex> state = %{
+          u1: %{metas: [%{id: 1, phx_ref: "1"}]},
+          u2: %{metas: [%{id: 2, phx_ref: "2"}]},
+          u3: %{metas: [%{id: 3, phx_ref: "3"}]}
+        }
+      iex> diff = %{
+        joins: %{u1: %{metas: [%{id: 1, phx_ref: "1.2"}]}},
+        leaves: %{u2: %{metas: [%{id: 2, phx_ref: "2"}]}}
+      }
+      iex> Phoenix.Presence.sync_diff(state, diff)
+      %{
+        u1: %{metas: [%{id: 1, phx_ref: "1"}, %{id: 1, phx_ref: "1.2"}]},
+        u3: %{metas: [%{id: 3, phx_ref: "3"}]}
+      }
+  """
+
+  def sync_diff(state, %{joins: joins, leaves: leaves}) do
+    Enum.reduce(leaves, state, fn leave, new_state ->
+      delete_meta_or_presence(new_state, leave)
+    end)
+    |> Map.merge(joins, fn _k, state_presence, joined_presence ->
+      %{metas: add_or_replace_metas(state_presence.metas, joined_presence.metas)}
+    end)
+  end
+
+  defp delete_meta_or_presence(state, {key, %{metas: left_metas}}) do
+    state_metas = Map.get(state, key, %{}) |> Map.get(:metas, [])
```

Comment:
@josevalim you mentioned _"discard the leaves of entries we don’t know"_

That happens in this function.

This will be an empty list`[]` if we receive such a leave.

So we will perform two operations on it
```elixir
# 1. Try to remove it from an empty list
[] -- [%{meta: :we_dont_know}]

# 2. Try to delete it from the state Map.
Map.delete(state, :key_we_dont_know)
```

Should I do some kind of early return (probably another function with a pattern match) if we do not know the meta to make this more performant? Or is removing from an empty list and deleting non-existent keys from a Map already very performant?